### ISSUE TITLE
TPU nan issue

### ISSUE THREAD

**Issue by khangtruong2252314** [tensorflow/tensorflow#86953]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

tf 2.18

### Custom code

Yes

### OS platform and distribution

Linux (google colab default)

### Current behavior?

In the 2.18.0 Tensorflow version, occur only when using TPU:
- At the middle of any epoch, loss turns out to be nan for the rest of the epoch 

### Standalone code to reproduce the issue

```shell
Shortest way: connect and connect to the colab tutorial on TPU, i.e. https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/guide/tpu.ipynb#scrollTo=Tce3stUlHN0L
If the issue hasn't been fixed, the training loop will produce nan loss
```

### Relevant log output

```shell
Epoch 1/5
300/300 ━━━━━━━━━━━━━━━━━━━━ 22s 47ms/step - loss: nan - sparse_categorical_accuracy: nan - val_loss: nan - val_sparse_categorical_accuracy: nan
Epoch 2/5
300/300 ━━━━━━━━━━━━━━━━━━━━ 9s 30ms/step - loss: nan - sparse_categorical_accuracy: nan - val_loss: nan - val_sparse_categorical_accuracy: nan
Epoch 3/5
300/300 ━━━━━━━━━━━━━━━━━━━━ 9s 30ms/step - loss: nan - sparse_categorical_accuracy: nan - val_loss: nan - val_sparse_categorical_accuracy: nan
Epoch 4/5
231/300 ━━━━━━━━━━━━━━━━━━━━ 1s 23ms/step - loss: nan - sparse_categorical_accuracy: nan
```

**Comment by Venkat6871** [comment]
Hi **@khangtruong2252314** ,

Apologies for the delay, and thanks for raising your concern here. I tried running your code on [Colab](https://colab.sandbox.google.com/gist/Venkat6871/6863307d5012a33382e72101baf5a959/86953_tpu_.ipynb) and encountered the same issue. The main cause is the Keras version. From TensorFlow 2.16.0 onwards, Keras 3 is installed by default. If you use Keras 2, the issue is resolved. To install Keras 2 manually, follow these steps:

```
!pip install tf-keras

import tf_keras as keras
```
I tried this in Colab, and it is working fine for me. I am providing a [gist](https://colab.sandbox.google.com/gist/Venkat6871/4519e726e86671e4188dd1b9ad19e247/86953_tf_2-18-0-v-tpu.ipynb) here for your reference.

Thank you!

**Comment by khangtruong2252314** [comment]
Thank you very much, but I don't think focus on the difference between Keras 3 and Keras 2 would find the bug, actually, since when I ran Tensorflow 2.17.1 last week, the issue didn't exist. But when the colab update TF version to 2.18, this nan error cause in all notebooks I found. One more thing is, despite the tf_keras can run in the notebook I provided, my project (I did not provide here) cause crashes which did not work. I believe focusing on TF 2.17 and TF 2.18 alone could be more promising.

**Comment by IrinaLuminesk** [comment]
Did you find a solution for this problem? I have tried switching to the GPU, but this caused the training time to be very slow.

**Comment by khangtruong2252314** [comment]
Ah, please don’t make the nan loss bother you. I still get the same test metrics as when the issue didn’t happen (which indicate that the TPU did normally in the background but failed to print out). So the only problem you may encounter here is that you cannot see the metrics during training time, which is left unsolved. And perhaps you cannot use the callbacks during this time also. Anyway, you can still use TPU with some disadvantages.

**Comment by IrinaLuminesk** [comment]
@khangtruong2252314 I don't think it's a print out bug. If it were, you would still be able to access all the metrics using the history:
`
history = model.fit(training_dataset,
                    validation_data=validation_dataset,
                    steps_per_epoch=train_steps,
                    epochs=EPOCHS,
                    verbose=1)
train_loss = history.history['loss']
train_accuracy = history.history['sparse_categorical_accuracy']
val_loss = history.history['val_loss']
val_accuracy = history.history['val_sparse_categorical_accuracy']`. 
However, since the NaN bug you mentioned is still present, I can't access any of the training metrics, such as training loss and training accuracy. This is a huge problem for me because I use these metrics to publish articles about deep learning models."

**Comment by khangtruong2252314** [comment]
I suggest you have a testing phase after the training process to extract such metrics. I agree it is not printing issue, but the learning still occurs, which makes me guess that the TPU devices failed to transfer data back to CPU and directly lead to failure in any CPU related tasks such as history.
About the test phase, in my case, I switch to GPU and extract those output from the model which weight is stored on each epoch that I trained earlier in the training phase (well, I use Save model callback on every epoch and set Max epoch a fixed large number and then evaluate all the weights).

**Comment by kunalkukreja21** [comment]
I'm also facing a similar issue on multi-host TPUs. Issue does not happen on single-host TPUs. I'm using TPUv6e.

**Comment by IrinaLuminesk** [comment]
@kunalkukreja21 Did you initialized your TPUs using `tf.distribute.cluster_resolver.TPUClusterResolver` ?

**Comment by kunalkukreja21** [comment]
@IrinaLuminesk yes!

**Comment by IrinaLuminesk** [comment]
@kunalkukreja21 Well, that's strange. I am still encountering the NaN error. How did you use single-host TPUs, like you mentioned?

**Comment by van-flow** [comment]
I get this issue when using v2-8 tpu, but not when I switch to v5e-1 or v6e-1 tpus on Colab.
Reducing the number of batches to 32 or below also resolves the issue.

**Comment by Venkat6871** [comment]
Hi @khangtruong2252314, @IrinaLuminesk, @kunalkukreja21, @van-flow,

Thank you all for the additional details and for sharing your observations. From the discussion so far, multiple users have reported reproducing similar behavior on TPUs with TensorFlow 2.18, while others have reported that the issue does not occur with `tf-keras` or on newer TPU configurations.

Could anyone still experiencing this issue try reproducing it with the latest `tf-nightly` build? If possible, please also include:
* the TPU type (for example, v2-8, v3-8, v5e, or v6e),
* whether the issue occurs on single-host or multi-host TPUs, and
* a minimal reproducible example, if different from the official Colab tutorial.

This additional information will help determine whether the issue is specific to a particular TPU configuration or if it has already been addressed in newer builds.

Thank you!
